### DISCUSSION TITLE
Encoder16 bit , i2s 32 bit

### DISCUSSION THREAD

**Discussion by aqildad-create** [pschatzmann/arduino-audio-tools#727] [Q&A] [answered]
I have INMP441 which give 32 bit data and want to pas this to encoder , and as per my knowledge encoder support 16 bit data , 

further I am facing difficulty in converting 32 bit data to 16 bit data .  

For this  i   remove upper 16 bit from array and lower 16 bit will my required data , is my observation  correct ??

 below is my Pseudocode .

 

File dataFile = SD.open("example.wav", FILE_WRITE); 
WAVEncoder encoder(dataFile);
 .
.
.
static uint8_t buffer32[500];
static uint8_t buffer16[250];

if (i2s.available()>0){
            int len = i2s.readBytes(buffer32, 500);
           byte passData =0;
           int index16=0;
            for ( int a=0 ;a<500; a++)
            {      
           passData =passData +1;
          if (passData <= 2 )
              buffer16[ndex16]=buffer32[a] ;
              index16=index16+1;
                       }
           if( passData ==4 )  passData =0;

            encoder.write(buffer16, index16);
        }  

            
-----------32------------||--------------32------------||
| 8bit | 8bit | 8bit | 8bit  || 8bit| 8bit|8bit| 8bit|
------------------------- ||------------------------------||
---16------||------16-----||
| 8bit | 8bit || 8bit | 8bit  || 
------------------------- ||

**Comment by pschatzmann** [comment, marked as answer]
Why don't just just use the [FormatConverterStream](https://github.com/pschatzmann/arduino-audio-tools/wiki/Converting-the-Data-Format) ?
If you want to program it yourself, I suggest to do some casts, then you can do regular calculations. The input it int32_t* and the result should end in int16_t*. 

If you really want to wast some memory your buffer32 should be int32_t and your buffer16 should be int16_t.

**Reply by aqildad-create** [reply]
without casting even i try this . . file was written but with lot of noise , 

using your class **coper.copy**  it record perfect with 32bit . with 16 bit conversion its not working 

if (i2s.available()>0){
                      int len = i2s.readBytes(buffer32, 1024);
                      encoder.write(buffer32, len);
}
.
.
file.close();

still working on FormatConverterStream with espnow example ..but no success yet

**Comment by aqildad-create** [comment]
I try to use FormatConverterStream  but  i am missing some thing , because the i cant listen audio  

Below  i try to use FormatConverterStream 

https://github.com/pschatzmann/arduino-audio-tools/discussions/719#discussioncomment-5449463

**Comment by pschatzmann** [comment]
Just did a test, and I confirm that there seems to be an issue. However the following is working w/o any issues
```C++

#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"

SineWaveGenerator<int32_t> sine_wave;                   // subclass of SoundGenerator with max amplitude of 32000
GeneratedSoundStream<int32_t> in_stream(sine_wave);     // Stream generated from sine wave
//CsvStream<int16_t> out(Serial); 
AudioKitStream out;
AudioInfo from(44100, 2, 32);                                           
AudioInfo to(44100, 2, 16);                                           
NumberFormatConverterStream conv(out);

StreamCopy copier(conv, in_stream);                     // copies sound to out

void setup(){
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);  

  sine_wave.begin(from, N_B4);
  in_stream.begin(from);
  conv.begin(from.bits_per_sample, to.bits_per_sample);

  auto cfg = out.defaultConfig(TX_MODE);
  cfg.copyFrom(to);
  out.begin(cfg);
}

void loop(){
    copier.copy();
}

```


You could also use the typed version by replacing

```C++
NumberFormatConverterStreamT<int32_t, int16_t> conv(out);
conv.begin();

```

**Reply by pschatzmann** [reply]
The bug in the FormatConverterStream has been corrected. 
There was one method, which was using the parameters in the wrong order...

**Comment by aqildad-create** [comment]
with typed  version . but no success , now log is also different , 

#include "AudioTools.h"
#include "AudioLibs/Communication.h"
#include "AudioCodecs/CodecSBC.h"
I2SStream i2sStream; // Access I2S as stream
uint16_t sample_rate = 44100;
//AudioInfo to(44100, 2, 16);
//FormatConverterStream conv(i2sStream);
NumberFormatConverterStreamT<int32_t, int16_t> conv(i2sStream);

ESPNowStream now;
EncodedAudioStream encoder(&now, new SBCEncoder()); // encode and write to ESP-now

StreamCopy copier(encoder, conv);

const char *peers[] = {"0C:B8:15:C3:40:68"};
void setup() {
Serial.begin(115200);
AudioLogger::instance().begin(Serial, AudioLogger::Info);

auto cfg = now.defaultConfig();
cfg.mac_address = "94:E6:86:04:89:44";

now.begin(cfg);
now.addPeers(peers);

// start i2s input with default configuration
Serial.println("starting I2S...");
auto config = i2sStream.defaultConfig(RX_MODE);
config.i2s_format = I2S_STD_FORMAT; // if quality is bad change to I2S_LSB_FORMAT #23
config.sample_rate = sample_rate;//22050;
config.channels = 2;
config.bits_per_sample =32;
i2sStream.begin(config);
Serial.println("I2S started");
conv.begin();//conv.begin(config, to);
// start encoder
encoder.begin();

Serial.println("Sender started...");
}

void loop() {
copier.copy();
}

[I] Communication.h : 107 - setting mac 94:E6:86:04:89:44

mac: 94:E6:86:04:89:44
[I] Communication.h : 310 - esp_now_init: 94:E6:86:04:89:44
[I] Communication.h : 161 - addPeer: 0c:b8:15:c3:40:68
starting I2S...
[I] I2SConfig.h : 106 - rx/tx mode: RX_MODE
[I] I2SConfig.h : 107 - port_no: 0
[I] I2SConfig.h : 108 - is_master: Master
[I] I2SConfig.h : 109 - sample rate: 44100
[I] I2SConfig.h : 110 - bits per sample: 32
[I] I2SConfig.h : 111 - number of channels: 2
[I] I2SConfig.h : 112 - i2s_format: I2S_STD_FORMAT
[I] I2SConfig.h : 114 - auto_clear: false
[I] I2SConfig.h : 116 - use_apll: true
[I] I2SConfig.h : 121 - buffer_count:6
[I] I2SConfig.h : 122 - buffer_size:512
[I] I2SConfig.h : 126 - pin_bck: 14
[I] I2SConfig.h : 128 - pin_ws: 15
[I] I2SConfig.h : 130 - pin_data: 32
I2S started
[I] CodecSBC.h : 247 - virtual void audio_tools::SBCEncoder::begin()
[E] CodecSBC.h : 330 - Invalid sample_rate: 0
Sender started...
[E] CodecSBC.h : 273 - inactive
[E] CodecSBC.h : 273 - inactive
[I] AudioCopy.h : 292 - try write - 2 (open 250 bytes) 
[E] CodecSBC.h : 273 - inactive
[I] AudioCopy.h : 292 - try write - 3 (open 250 bytes) 
[E] CodecSBC.h : 273 - inactive
[I] AudioCopy.h : 292 - try write - 4 (open 250 bytes) 
[E] CodecSBC.h : 273 - inactive
[I] AudioCopy.h : 292 - try write - 5 (open 250 bytes) 
[E] CodecSBC.h : 273 - inactive
[I] AudioCopy.h : 292 - try write - 6 (open 250 bytes) 
[E] CodecSBC.h : 273 - inactive
[I] AudioCopy.h : 292 - try write - 7 (open 250 bytes) 
[E] CodecSBC.h : 273 - inactive
[I] AudioCopy.h : 292 - try write - 8 (open 250 bytes) 
[E] CodecSBC.h : 273 - inactive
[I] AudioCopy.h : 292 - try write - 9 (open 250 bytes) 
[E] CodecSBC.h : 273 - inactive
[I] AudioCopy.h : 292 - try write - 10 (open 250 bytes) 
[E] CodecSBC.h : 273 - inactive
[I] AudioCopy.h : 292 - try write - 11 (open 250 bytes) 
[E] CodecSBC.h : 273 - inactive
[I] AudioCopy.h : 292 - try write - 12 (open 250 bytes) 
[E] CodecSBC.h : 273 - inactive
[I] AudioCopy.h : 292 - try write - 13 (open 250 bytes) 
[E] CodecSBC.h : 273 - inactive
[I] AudioCopy.h : 292 - try write - 14 (open 250 bytes) 
[E] CodecSBC.h : 273 - inactive
[I] AudioCopy.h : 292 - try write - 15 (open 250 bytes) 
[E] CodecSBC.h : 273 - inactive
[I] AudioCopy.h : 292 - try write - 16 (open 250 bytes) 
[E] CodecSBC.h : 273 - inactive
[I] AudioCopy.h : 292 - try write - 17 (open 250 bytes) 
[E] CodecSBC.h : 273 - inactive
[I] AudioCopy.h : 292 - try write - 18 (open 250 bytes) 
[E] CodecSBC.h : 273 - inactive
[I] AudioCopy.h : 292 - try write - 19 (open 250 bytes) 
[E] CodecSBC.h : 273 - inactive
[I] AudioCopy.h : 292 - try write - 20 (open 250 bytes) 
[E] CodecSBC.h : 273 - inactive
[I] AudioCopy.h : 292 - try write - 21 (open 250 bytes) 
[E] CodecSBC.h : 273 - inactive
[E] AudioCopy.h : 286 - write to target has failed! (250 bytes)
[I] AudioCopy.h : 136 - StreamCopy::copy 250 -> 250 -> 0 bytes - in 22 hops
[E] CodecSBC.h : 273 - inactive
[E] CodecSBC.h : 273 - inactive

**Comment by pschatzmann** [comment]
[E] CodecSBC.h : 330 - Invalid sample_rate: 0
So why do you think that a sample rate of 0 would work ?

**Reply by aqildad-create** [reply]
How the sample rate become zero ?? see the mage  
![image](https://user-images.githubusercontent.com/65456998/228757595-56357e9a-96b3-47dc-9e58-45908a5933ca.png)

**Reply by pschatzmann** [reply]
You were defining it for the input side (I2S) but not for the output side (EncodedAudioStream or ESPNowStream)
I try to deduce it if ever possible, but in this case there is no link between the input and the output

**Reply by aqildad-create** [reply]
@pschatzmann  thanks for the consideration , 

yes sir  I agree with you ,

But Pls give a sample code which get i2s ( INMP441 )  data and send to espnow with SBC encoded  , i have try many attempts but failed all the time .
